### ISSUE TITLE
Conditional Focus (When Display-Capture Starts)

### ISSUE THREAD

**Issue by eladalon1983** [w3c/mediacapture-screen-share#190]
## Problem
When an application starts capturing a display-surface, the user agent faces a decision - should the captured display-surface be brought to the forefront, or should the capturing application retain focus.

The user agent is mostly agnostic of the nature of the capturing and captured applications, and is therefore ill-positioned to make an informed decision.

In contrast, the capturing application is familiar with its own properties, and is therefore better suited to make this decision. Moreover, by reading [displaySurface](https://www.w3.org/TR/screen-capture/#dom-mediatrackconstraintset-displaysurface) and/or using [Capture Handle](https://wicg.github.io/capture-handle/), the capturing application can learn about the captured display-surface, driving an even more informed decision.

## Sample Use Case
For **example**, a video conferencing application **may** wish to:
* Focus a captured application that users typically interact with during the call, like a text editor.
* Retain for itself focus when the captured display-surface is non-interactive content, like a video.
  * (Using [Capture Handle](https://wicg.github.io/capture-handle/), the capturing application may even allow the user to remotely start/pause the video.)

## Suggested Solution and Demo
* I have produced a [spec-draft](https://eladalon1983.github.io/conditional-focus/index.html) of my suggested solution.
* I have also implemented it behind a flag in Chrome and produced and a [demo](https://eladalon1983.github.io/conditional-focus/demo/).
  * Note that this demo works on Chrome Canary.
  * Further instructions are provided in the demo itself.

## Sample Code
```
const stream = await navigator.mediaDevices.getDisplayMedia();
track.focus(ShouldFocus(stream) ? "focus-captured-surface" : "no-focus-change");

function ShouldFocus(stream) {
  const [track] = stream.getVideoTracks();
  if (sampleUsesCaptureHandle) {
    // Assume logic discriminating focusability by origin,
    // for instance focusing anything except https://collaborator.com.
    const captureHandle = track.getCaptureHandle();
    return ShouldFocusOrigin(captureHandle && captureHandle.origin);
  } else {  // Assume Capture Handle is not a thing.
    // Assume the application is only interested in focusing tabs, not windows.
    return track.getSettings().displaySurface == 'browser';
  }
}
```

## Security Concerns
One noteworthy security concerns is that allowing switching focus at an arbitrary moment could allow clickjacking attacks. The [suggested spec](https://eladalon1983.github.io/conditional-focus/index.html) addresses this concern by limiting the time when focus-switching may be triggered/suppressed - the application may only decide about focus immediately[\*] upon the resolution of the `Promise<MediaStream>`. (See the [spec-draft](https://eladalon1983.github.io/conditional-focus/index.html) for more details about what "immediately" means and how I suggest various edge-cases be handled.)

**Comment by happylinks** [comment]
This looks amazing, been waiting for this for a long time! In tella.tv we have the following flow:
1. User goes to recorder
2. User selects a screen/window/tab they want to record
3. When picking window or tab, they get jumped to the window or tab, but need to go back to tella first to start the recording.
4. Start recording in tella
5. Countdown
6. Manually go to the tab/window they were recording.

If i'm looking at this proposal right, our new flow could be:
1. User goes to recorder
2. User selects a screen/window/tab they want to record (and does not get jumped to the window or tab)
3. User clicks start recording
4. Countdown
5. We fire .focus on the stream to bring them where they need to be automatically.

A much nicer flow. Would love to be in the Origin Trial when it's available :)

**Comment by happylinks** [comment]
Ah, reading the spec again, missed this point:
```If this call to focus() happens more than one second after the start of the capture, the user agent SHOULD have already made its own decision. The user agent SHOULD silently ignore this call focus().```

That means we would not be able to do "We fire .focus on the stream to bring them where they need to be automatically." after countdown.
Still better to keep them in Tella though! What is the reason for the 1 second if I may ask ?

**Comment by eladalon1983** [comment]
> What is the reason for the 1 second if I may ask ?

Please note that 1s is the hard-limit, and normally the window of opportunity to calling `focus()` would close even earlier - as soon as the microtask on which the `Promise` was resolved is completed. (The 1s hard-limit prevents the application from keeping that microtask running for N seconds and then switching focus.)

The rationale for these limitations is preventing attacks where the capturing application induces the user to click at it chosen location in the captured application by switching focus at just the right time. As a somewhat contrived example, imagine the capturing application presenting at _just_ the right location a button labelled "double-click here to confirm" and then switching focus after it registers the first click.

Are these concerns realistic? I'm not yet sure, but it should be easier to reach consensus if we err on the side of caution.

**Comment by happylinks** [comment]
Thanks for explaining! Preventing the focus will already help us a lot.

**Comment by eladalon1983** [comment]
@jan-ivar and @youennf, I'd love to carry on from where we stopped at the end of the WebRTC WG Interim meeting yesterday.

@youennf
IIUC, your main concern was around the 1s hard-limit, but I think that you later mentioned this was due to a misunderstanding? Could you PTAL at my [proposed spec](https://eladalon1983.github.io/conditional-focus/index.html) and let me know what you think? Please see this part:
> 3. If focus() was not called on MT, raise an InvalidStateError. Otherwise, proceed.
> 4. If this call to focus() happens more than one second after the start of the capture, the user agent SHOULD have already made its own decision. The user agent SHOULD silently ignore this call focus().

Please note that the intention here is that if focus() is **not** called by the time MT terminates, then the browser **immediately** switches focus - **without** a 1s delay. The 1s delay is a backup in case MT takes an inordinately long time to finish.

**Comment by eladalon1983** [comment]
@jan-ivar and @youennf, I'd love to carry on from where we stopped at the end of the WebRTC WG Interim meeting yesterday.

@youennf
IIUC, your main concern was around the 1s hard-limit, but I think that you later mentioned this was due to a misunderstanding? Could you PTAL at my [proposed spec](https://eladalon1983.github.io/conditional-focus/index.html) and let me know what you think? Please see this part:
> 3. If focus() was not called on MT, raise an InvalidStateError. Otherwise, proceed.
> 4. If this call to focus() happens more than one second after the start of the capture, the user agent SHOULD have already made its own decision. The user agent SHOULD silently ignore this call focus().

Please note that the intention here is that if focus() is **not** called by the time MT terminates, then the browser **immediately** switches focus - **without** a 1s delay. The 1s delay is a backup in case MT takes an inordinately long time to finish.

@jan-ivar
1. IIUC, your concerns where over subclassing MediaStreamTrack vs. expsosing an API as `navigaotr.mediaDevices.focus`. Could you please elaborate on why you think one is superior to the other?
2. I believe you also suggested scheduling a task to close the focus-window-of-opportunity. My concern here is that tasks are more likely than microtasks to experience long delays, and the hard-limit of 1s would be encountered more often than we'd want. (I think we'd want it as a security restriction that's hardly ever encountered by a non-malicious app.) Could you please elaborate your thoughts on this topic?

Please let me know if I've forgotten any other open issues.

**Comment by eladalon1983** [comment]
Btw, following the interim it became clearer to me which parts of my [proposed spec](https://eladalon1983.github.io/conditional-focus/index.html) were in need of clarification. I have made adjustments, mostly around (a) explaining the window of opportunity and (b) providing examples. PTAL.

**Comment by jan-ivar** [comment]
> 1. IIUC, your concerns where over subclassing MediaStreamTrack vs. expsosing an API as `navigaotr.mediaDevices.focus`. Could you please elaborate on why you think one is superior to the other?

A user's focus is a global property, so a per-track API [simply](https://w3ctag.github.io/design-principles/#simplicity) isn't necessary, and we can avoid the subclassing headache.

> I believe you also suggested scheduling a task to close the focus-window-of-opportunity. My concern here is that tasks are more likely than microtasks to experience long delays, and the hard-limit of 1s would be encountered more often than we'd want. (I think we'd want it as a security restriction that's hardly ever encountered by a non-malicious app.) Could you please elaborate your thoughts on this topic?

I think you're looking at this the wrong way. How does a tighter window of opportunity benefit slow apps?

What matters is the amount of code that runs up to the point of invoking the focus method. In your model, JS gets a whole second as long as it stays synchronous, but a single microtask switch (e.g. `async undefined`) and it's out.

That seems too harsh to me, making things harder on apps, not easier. It also seems biased against apps that use promises.

Microtask-switching is sometimes unavoidable when writing promise code (and arguably why microtasks exist — their whole point is that they're not tasks, and operate within a task). For example:
```js
async function getUserMedia(constraints) {
  const stream = await navigator.mediaDevices.getUserMedia(constraints);
  // store track ids in localStorage
  return stream;
}
```
This introduces a microtask switch on code calling this function as a drop-in for the real thing — but it's on the same task.🤔

From a security perspective, tasks and microtasks are equally vulnerable to busy-looping.

**Comment by jan-ivar** [comment]
@alvestrand asked me to outline the API shape I propose. I'd prefer:
```js
console.log(navigator.mediaDevices.displayMediaFocusMode); // "focus"

const stream = await navigator.mediaDevices.getDisplayMedia();
navigator.mediaDevices.displayMediaFocusMode = "no-focus"; // stream's target won't be focused
```
It's a simple read-write enum attribute with two values. It never changes value except when JS changes it.

UA reads it in a task queued from `getDisplayMedia`'s success task, to determine whether to focus what was just resolved.

The app can set it ahead of time as well, if it doesn't care what the user picks:
```js
navigator.mediaDevices.displayMediaFocusMode = "no-focus"; // stream's target won't be focused
const stream = await navigator.mediaDevices.getDisplayMedia();
```

**TL;DR:** it affects subsequent calls, _but_ if you set it immediately, it affects what just resolved as well.

**Comment by eladalon1983** [comment]
> `...navigator.mediaDevices.displayMediaFocusMode = "no-focus";`

I don't think this is going to "play nice" if the document is engaged in multiple concurrent captures.

> it affects subsequent calls, but if you set it immediately, it affects what just resolved as well

I don't think that's [simple](https://w3ctag.github.io/design-principles/#simplicity), because the association between what gets captured and what gets focused is not there, and concurrent captures become clouded by questions of timing.

> A user's focus is a global property, so a per-track API [simply](https://w3ctag.github.io/design-principles/#simplicity) isn't necessary, and we can avoid the subclassing headache.

How is `navigator.mediaDevices.someFocusApi` different than `FMST.focus()` with respect to "globality"?
`FMST.focus()` allows feature discovery; it'd be impossible for an application to attempt to call it on a non-supporting track. So that's a benefit of this approach. What benefits would the other approach confer?

> What matters is the amount of code that runs up to the point of invoking the focus method.

One does not need a lot of code to place a "Click here to win a million dollars" button and then guess how long to wait before switching focus to the other tab and getting the user to click something there. (Moreover, if you use tasks rather than microtasks, I think it's easier to wait for an on-over event, making guesses about when the click is coming easier. But I could be wrong there. Let's call this a secondary argument.)

I don't think we can measure and base a decision off of how much code runs, regardless of whether the window-of-opportunity is closed by a task or a microtask. Either could run an arbitrary amount of code. The user's reaction time, though...

> It also seems biased against apps that use promises.

Some of my best friends are apps that use Promises. I have found them to be capable of switching between using Promises and doing things synchronously, as the need arises.

**Comment by jan-ivar** [comment]
> I don't think this is going to "play nice" if the document is engaged in multiple concurrent captures.

It should: There's only one user, who can only accept one prompt at a time. The association is with the prompt — the user interaction — not the track.

Example: Imagine a hypothetical `navigator.mediaDevices.displayMediaFocus(track)` method that must be called between a pair of tasks queued sequentially by the UA, where the first resolves with a new screen-sharing track: Context tells us what to focus, and the `track` argument is entirely redundant.

> `FMST.focus()` allows feature discovery

This should suffice:
```js
if ('displayMediaFocusMode' in navigator.mediaDevices) { /* UA feature detected */ }
```

All browsers today automatically focus all focusable surfaces (`"monitor"` isn't focusable), so the default is `"focus"`.

I'd recommend we strongly encourage UAs to allow apps to override this automatic focus for all focusable surfaces.

This removes any need for per-surface-type detection of policies.

> One does not need a lot of code to place a "Click here to win a million dollars" button and then guess how long to wait before switching focus to the other tab and getting the user to click something there. (Moreover, if you use tasks rather than microtasks, I think it's easier to wait for an on-over event, making guesses about when the click is coming easier.

We discussed offline that gDM can queue two tasks A and B, guaranteeing no other task runs in between, so events aren't an issue. The promise is resolved in A, creating microtask A₁ where JS runs. Your proposal is to limit calls to between A₁-A₂, whereas I'm saying A₁-B is fine and superior as it avoids tripping up promise-use. The 1 second timeout is orthogonal.

**Comment by youennf** [comment]
I also prefer an attribute over an explicit call for each track.
This allows an application to set its default value easily (never focus for instance) while providing enough flexibility for web applications that want per-capture decisions.

As of event task vs. micro task, there are preexisting examples.
For instance, the fetch event is fired and some properties are read synchronously after the fetch event listeners, thus before the microtasks that would have been scheduled during the fetch event listeners.

The main advantage I see is that it is a clear web developer contract: the boundary to call focus is the next await.
I believe this approach also allows shiming since the read-focus callback would be executed after the already registered promise resolution callback.

From a pure spec/developer point of view, it should be also easier to specify and implement: queue a task where the promise is resolved and a promise resolution callback to read the value and apply focus is registered.
If we would like to leave the whole task to change the focus value, we might actually not want to read the value in the next task but at the point in the current task where the microtask queue gets emptied after the task steps are executed.
I am not sure how doable that is right now in terms of spec editing.

**Comment by jan-ivar** [comment]
> I also prefer an attribute over an explicit call for each track.

👍

> As of event task vs. micro task, there are preexisting examples.
> For instance, the fetch event is fired and some properties are read synchronously after the fetch event listeners, thus before the microtasks that would have been scheduled during the fetch event listeners.

Event listeners must respond synchronously to prevent bubbling/avoid default for historical reasons.

Those "properties" appear to be [fetch](https://w3c.github.io/ServiceWorker/#on-fetch-request-algorithm)  trying to _support_ promise code with [`event.respondWith(async () =>  {}`)](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent#examples). So my takeaway from fetch is we should make affordances for promise use. (_"[[In fetch] we use promises heavily](https://w3c.github.io/ServiceWorker/#execution-context)"_). 

> I believe this approach also allows shiming since the read-focus callback would be executed after the already registered promise resolution callback.

I'll try to explain the shimming problem in more detail: Say a JS library (e.g. [adapter.js](https://github.com/webrtcHacks/adapter/blob/ba8d60c751d33b4c4eefdf2c30dbe1ccb33a3f4e/src/js/firefox/getdisplaymedia.js#L11)) needs to shim gDM for some reason:
```js
const nativeGDM =  window.navigator.mediaDevices.getDisplayMedia;

window.navigator.mediaDevices.getDisplayMedia = async function getDisplayMedia(constraints) {
  // Here we're on Task A
  const stream = await nativeGDM.apply(this, arguments);
  // Here we're on Task B, Microtask B₁
  stream.newFeatureX = 3;
  return stream; // the implicit promise returned by this async function is resolved with stream
}
```
The shim is careful not to await anything else, yet a microtask checkpoint is unavoidable, because the promise returned by the shim is not the same as the one from `nativeGDM`. Thus the JS library will unintentionally break all existing code attempting to suppress focus. E.g.:
```js
// Here we're on Task A
const stream = await navigator.mediaDevices.getDisplayMedia({video: true});
// Here we'd be on Task B, Microtask B₁ without the shim, but with the shim we're on Microtask B₂ instead
navigator.mediaDevices.displayMediaFocusMode = "no-focus"; // focuses with the shim but not without!
```

**Comment by eladalon1983** [comment]
### Global or per-surface controls

> (@youennf:) I also prefer an attribute over an explicit call for each track.

These are two distinct preferences:
1. Global or per-surface. (Discussed in this section.)
2. Attribute or method. (Discussed in the next section.)

> (@jan-ivar:) There's only one user, who can only accept one prompt at a time.

The browser can operate in modes which skip the prompt. Mechanisms to trigger these include extensions, enterprise policies and command-line arguments. The spec may be agnostic of these, but the fact remains that the user can accept multiple different captures within a very short span of time.

At any rate, if the application fires off two calls to getDisplayMedia and wants to focus exactly one of these, then it's a lot **more ergonomic** to call focus() on the right track, than to to manipulate a global attribute at just the right time, ensuring it's the intended value when the UA reads it for the one display-surface and the other value when the UA reads it for another display-surface. It requires of the Web-developer much more in-depth understanding.

### Method vs. Attribute

Assume, for the sake of argument, that my previous section convinced you to use per-surface controls. Do we want a method or attribute then?

An application that can read the value may just as well set its own preferred value. An attribute for `focus` would make sense if global, but not if per-surface.

**However**, before even calling `getDisplayMedia`, the application might already wish to know whether it can influence the decision. I would not object to adding a global attribute that reads (and potentially writes?) the default behavior - that which applies if the per-surface API is not invoked. The presence of this global attribute also informs the application that the per-surface control will be exposed if the user chooses to share a focusable surface.

Btw, one challenge with writable attributes is that Web-developers would not as readily expect setting of attributes to potentially raise an exception.

### Subclassing MediaStreamTrack

I think we have seen multiple cases where subclassing MediaStreamTrack would have confered benefits, but each time a discussion arose over whether it's enough to sub-class just for that. The results of having everything on MediaStreamTrack is sub-optimal. Some immediate beneficiary APIs of a decision to sub-class would be:
* The focus API - only focusable surfaces (browser-surfaces and windows; in the future potentially applications and/or isolated-browser-surfaces).
* Capture Handle - only browser-surfaces (maybe more in the future).
* Cropping - only self-capture (SelfCaptureMST -> BrowserSurfaceMST -> FocusableMST -> DisplayCaptureMST -> MST).

IMHO, this list is sufficiently long and the benefits are sufficient. When someone calls `getViewportMedia()`, the result is inherently different than when someone calls `getUserMedia()`, and it makes sense for the APIs exposed to reflect that.

### Tasks vs. Microtasks

IIANM, the only argument for tasks is that they are shim-friendly. (Please correct me if I'm wrong.)

An argument against tasks is that in addition to shimming, it allows an application to `await somePromise`. This is an **anti-pattern**, as the results would be flaky. If `somePromise` is already resolved, it would work; if it will only by resolved by a later task, it would not work.

This trade-off is easy to reason about (IMHO) because **we can have both**. If we use microtasks, shimming is possible with an adapter:
```js
const nativeGDM =  window.navigator.mediaDevices.getDisplayMedia;

function focusCallback(stream) {
  // Return "no-focus-change" or "focus-captured-surface"
}

window.navigator.mediaDevices.getDisplayMedia = async function getDisplayMedia(constraints, focusCallback) {
  const stream = await nativeGDM.apply(this, arguments);
  const [track] = stream.getVideoTracks();
  if (!!focusCallback && !!track.focus) {
    const shouldFocus = focusCallback(stream);
    track.focus(shouldFocus);
  }
  return stream;
}
```

The code outside the shim just plugs their callback. Note that there are natural limits on what the app can do anyway until the window-of-opportunity closes, so I expect the code would easily and naturally fit inside of a synchronous callback.

**Comment by youennf** [comment]
> The shim is careful not to await anything else, yet a microtask checkpoint is unavoidable

It is avoidable by having the shim returning the native promise returned by getDisplayMedia.
The only downside is that you cannot change the object being resolved, but you can update the object itself: add tracks, remove tracks...

Looking at the HTML event loop, we can also see that after the micro task checkpoint, there are additional tasks that need to be done, some of which might trigger firing events.

> These are two distinct preferences:
> 
> 1. Global or per-surface. (Discussed in this section.)
> 2. Attribute or method. (Discussed in the next section.)

I was meaning a global attribute, somewhere in navigator.
With the current focus API design, in particular the restriction to read the focus-or-not value synchronously in the task resolving the promise, I do not see a need to have a per-track method.

Let's say the application wants a behavior for all getDisplayMedia calls. It will set the attribute once.
Let's say the application wants to select the behavior for each getDisplayMedia call, it will update the attribute value each time within the promise resolution callback. 

A method is less intuitive to me.

> one challenge with writable attributes is that Web-developers would not as readily expect setting of attributes to potentially raise an exception.

I do not think the plan for setting this attribute is to raise an exception.

**Comment by eladalon1983** [comment]
> Let's say the application wants to select the behavior for each getDisplayMedia call, it will update the attribute value each time within the promise resolution callback.

A Web-developer would need non-trivial understanding of the feature in order to become convinced that there'd be no carry-over effect of focusing the wrong surface. They might explain it to themselves as "only the last capture gets focus or not" - but then, why does it sometimes not seem to work? There is no exception raised when they set the new value too late (e.g. `await` on an yet-unresolved-Promise). With a method, there would be, and that exception would not be unexpected.

**Comment by jan-ivar** [comment]
> It is avoidable by having the shim returning the native promise returned by getDisplayMedia.
> The only downside is that you cannot change the object being resolved, but you can update the object itself: add tracks, remove tracks...

@youennf How? Please be specific.

> > one challenge with writable attributes is that Web-developers would not as readily expect setting of attributes to potentially raise an exception.
> 
> I do not think the plan for setting this attribute is to raise an exception.

Agreed (not beyond WebIDL's `TypeError` for invalid enum values).

**Comment by eladalon1983** [comment]
> > I do not think the plan for setting this attribute is to raise an exception.
> 
> Agreed (not beyond WebIDL's `TypeError` for invalid enum values).

Contrast:

#### 1.
```js
const stream = await navigator.mediaDevices.getDisplayMedia();
const [track] = stream.getVideoTracks();
await someOtherPromisesThatResolvesMuchLater;
track.focus("no-focus-change");
```

#### 2.
```js
const stream = await navigator.mediaDevices.getDisplayMedia();
const [track] = stream.getVideoTracks();
await someOtherPromisesThatResolvesMuchLater;
navigator.mediaDevices.focusPolicy = "no-focus-change";
```

#### 3.
```js
const stream = await navigator.mediaDevices.getDisplayMedia();
const [track] = stream.getVideoTracks();
await immediatelyResolvedPromise;
navigator.mediaDevices.focusPolicy = "no-focus-change";
const otherStream = await navigator.mediaDevices.getDisplayMedia();
```


* The first can throw an exception that explains to Web-developers why what they're doing is not having the desired effect. 
* The second has to accommodate the option that the Web-developer actually intended to set the *new* policy, going forward, completely independently of the call to gDM that has already resolved. The UA cannot raise an exception, because no error was made - but we human beings can tell there's a good chance it's a bug in the Web-application.
* The third affects the gDM call that just resolved. Maybe it's what the developer wanted; or maybe they actually intended to only affect `otherStream`, and this is actually the inverse bug of the second example. Who knows?!

Only the first option produces consistent results (and raises a clear exception when used incorrectly).

**Comment by jan-ivar** [comment]
> The browser can operate in modes which skip the prompt.

That is out of scope for this working group.

> The spec may be agnostic of these, but the fact remains that the user can accept multiple different captures within a very short span of time.

@eladalon1983 Not a problem. JS is single-threaded, and [getDisplayMedia](https://w3c.github.io/mediacapture-screen-share/#dom-mediadevices-getdisplaymedia) implicitly queues a task to resolve a promise from in parallel steps. My proposal would be to queue two tasks, guaranteed to happen in succession. There is no "span of time" short enough for two `getDisplayMedia` calls resolving to be a problem.

>  it's a lot more ergonomic to call focus() on the right track, than to to manipulate a global attribute at just the right time,

You have to call focus() at just the right time as well. Also, if ergonomics is the issue, why isn't it `stream.focus()`?

**Comment by eladalon1983** [comment]
> > The browser can operate in modes which skip the prompt.
> 
> That is out of scope for this working group.

I have foreseen this response and added the text which you quoted immediately below (agnosticism etc.). The topic is not whether we can handle it spec-wise, but rather whether it produces ergonomic results for the application. Consider:

```js
const stream1 = await navigator.mediaDevices.getDisplayMedia();
doSomething();  // Maybe ends the task, maybe doesn't.
navigator.mediaDevices.focusPolicy = "no-focus"; // Who is affected? s1? s2? Both? Neither?
doSomethingElse();  // Maybe ends the task, maybe doesn't.
const stream2 = await navigator.mediaDevices.getDisplayMedia();
```

The global-attribute API produces code which is hard to reason about. Does it affect `stream1`? Does it affect `stream2`? Both? Neither? If flaky for either - what is the bug? It's definitely not simple.

> Not a problem. JS is single-threaded,

My assertion is **not** that it's impossible to spec this properly. I argue that the result is not ergonomic and not simple.

> Also, if ergonomics is the issue, why isn't it `stream.focus()`?

Well, if you think ergonomics would be improved by moving `focus()` to `MediaStream`, we can discuss. My motivation for **not** putting it there was that MediaStreams can get tracks added/removed, and then `focus()` becomes... un-simple.

**Comment by youennf** [comment]
> @youennf How? Please be specific.

```
const nativeGDM =  window.navigator.mediaDevices.getDisplayMedia;
window.navigator.mediaDevices.getDisplayMedia = function getDisplayMedia(constraints) {
  const promise = nativeGDM.apply(this, arguments);
  promise.then(stream => {
    stream.newFeatureX = 3;
  });
  return promise;
}
```

**Comment by jan-ivar** [comment]
> Only the first option produces consistent results (and raises a clear exception when used incorrectly).

Of the three listed, which is a false choice. We don't need to subclass anything to throw. See below:

> I would not object to adding a global attribute that reads (and potentially writes?) the default behavior ...

So you'd introduce an attribute *and* a method?

Contrast:
1. ```js
   navigator.mediaDevices.defaultFocusPolicy = "focus-change";
   const stream = await navigator.mediaDevices.getDisplayMedia();
   await someOtherPromisesThatResolvesMuchLater;
   navigator.mediaDevices.focus("no-focus-change"); // throws InvalidStateError, focus already happened
   ```
vs.

2. ```js
   navigator.mediaDevices.focusPolicy = "focus-change";
   const stream = await navigator.mediaDevices.getDisplayMedia();
   await someOtherPromisesThatResolvesMuchLater;
   navigator.mediaDevices.focusPolicy = "no-focus-change"; // Doesn't throw, focus already happened
   ```

**Comment by jan-ivar** [comment]
> ```
> const nativeGDM =  window.navigator.mediaDevices.getDisplayMedia;
> window.navigator.mediaDevices.getDisplayMedia = function getDisplayMedia(constraints) {
>   const promise = nativeGDM.apply(this, arguments);
>   promise.then(stream => {
>     stream.newFeatureX = 3;
>   });
>   return promise;
> }
> ```

@youennf Sorry, but each `then` on the same promise introduces a microtask checkpoint.

**Comment by eladalon1983** [comment]
> So you'd introduce an attribute _and_ a method?

No. We're not on the same page. Maybe a quick summary of my position would help. I am **only** interested in a per-track method, but I am analyzing all options in my attempt to convince you that the option I favor is superior.

When I discuss inability to raise an exception, I do **not** mean it's because the attribute is an attribute. Rather, the problem arises because that global-attribute API is **always** valid to manipulate. Namely, an application could always be setting the value for the **next** gDM to follow.

I claim that inability to reliably raise an exception is a drawback. It leaves Web-developers in the dark when their applications behave inconsistently. It fails silently. That's a problem.

**Comment by jan-ivar** [comment]
> I am only interested in a per-track method,

Why, when it is less ergonomic than `stream.focus()`? 

> Rather, the problem arises because that global-attribute API is always valid to manipulate.

I showed an [option 1](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-928095773) above that would throw outside the same envelope, to show there's no need to subclass MST here.

From my end, the conversation has led to the options I show (1 and 2) which I think capture the remaining open question:

Would we rather offer JS a default value, or a clear exception when JS misses an (obvious) time window, or both (at the cost of double the API surface)?

**Comment by eladalon1983** [comment]
> > I am only interested in a per-track method,
> 
> Why, when it is less ergonomic than `stream.focus()`?

1. I have [answered](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-928092589) that question. Please see the bottom of my comment there.
2. Citation needed for one being more ergonomic than the other. (One being less code is not such a significant change in ergonomics. The ergonomics I refer to when I propose `MST.focus` are clarity, not brevity.)
3. I remain (cf. [comment](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-928092589)) open  to discussing a per-stream method . The context in which I mentioned "only per-track" was when you asked if I was suggesting ["an attribute _and_ a method"](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-928095773). It was not a blanket "only" for anything one could conceive of. It was a clarification that I was not at the time proposing two APIs, as you seemed to think. 

> I showed an option 1 above that would throw outside the same envelope, to show there's no need to subclass MST here.
> 
> From my end, the conversation has led to the options I show (1 and 2) which I think capture the remaining open question:

The original motivation I heard for using a global attribute, was that it allows to:
1. Probe the UA's capability before calling gDM.
2. Set a global default that would apply when gDM is called in the future.

So:
* I agree that (1) is achieved by the global attribute.
* I claimed that (2) loses the ability to raise instructive exceptions.

Do you now propose a version of the global that **does** raise exceptions? If so - that loses benefit (2). The UA cannot simultaneously, with a single API, allow the Web-developer to influence the future as well as warn them when they're too late to influence the past.

So my question is - which API are you now proposing, and what are its benefits over `MST.focus`?

**Comment by jan-ivar** [comment]
> ```js
> doSomething();  // Maybe ends the task, mabye doesn't.
> ```
Synchronous functions don't end a task, because JS is run to completion. Did you mean `await doSomething()` ?

> The global-attribute API produces code which is hard to reason about. Does it affect `stream1`? Does it affect `stream2`? Both? Neither? If flaky for either - what is the bug? It's definitely not simple.

I claim it is super simple, outside of contrived examples. Here's all users need to know:
 1. Await `navigator.mediaDevices.getDisplayMedia()`
 2. Set `navigator.mediaDevices.focusPolicy` before you do anything else.

All promise composition is allowed, including shims, `try`/`catch()`, breaking things into async sub-functions, `Promise.resolve(p)` and `Promise.race(p, timeout)`. This I claim is [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

If for some reason, you wish to explicitly _not_ affect the just resolved method, then `await new Promise(r => setTimeout(r))`

**Comment by jan-ivar** [comment]
Given how simple this is, I personally wouldn't bother with an exception.

**Comment by jan-ivar** [comment]
> All promise composition is allowed, including shims, `try`/`catch()`, breaking things into async sub-functions, `Promise.resolve(p)` and `Promise.race(p, timeout)`. This I claim is POLA.

In contrast, here's a list of benign things that'd cause action-at-a-distance focus-failure if we go with a microtask cutoff:
1. getDisplayMedia [shimming](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-926852042) with any synchronous post-success code at all
2. Promise-branching like @youennf's getDisplayMedia [shimming workaround](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-928099217) won't work either
3. Putting gDM into a [sub function](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-925408154) with any synchronous post-success code at all
4. Try-catch in a sub-function or shim
    ```js
   async function getDisplayMedia(c) {
      try {
       await navigator.mediaDevices.getDisplayMedia(c); // 💥 microtask checkpoint
     } catch (e) {
       log.addTo(e); // log all gDM errors
       throw e;
     }
   }
   ```
5. Promise.resolve()
   ```js
   function foo(mediaDevices) {
     return Promise.resolve(mediaDevices?.getDisplayMedia(c));  // 💥 microtask checkpoint
   }
   ```
6. Application timeout
   ```js
   const wait = ms => new Promise(r => setTimeout(r, ms));

   function timeout(p, ms) {
     return Promise.race([p, wait(ms)]);  // 💥 microtask checkpoint
   }
   await timeout(navigator.mediaDevices.getDisplayMedia(c), 10000); // 10s timeout
   ```

These are all things people expect to work when writing promise code, and there's no reason it should break some new API.

**Comment by youennf** [comment]
> @youennf Sorry, but each `then` on the same promise introduces a microtask checkpoint.

Each then returns a new promise, but this is fine.
The idea is for the microtask used by User Agent to check for focus-or-not to be executed within the same event loop task but after the JS shim getDisplayMedia promise resolution callbacks.
I think this is doable with the updated shim I provided, can you explain why it would not work?
I tried to express my understanding/POC in https://jsfiddle.net/f0uqx8jg/.

**Comment by alvestrand** [comment]
stream.focus() is a red herring. A stream is not a reference to a focusable object; the fact that a stream returned from GetDisplayMedia is guaranteed to have just one video track is not a property of MediaStream.

**Comment by eladalon1983** [comment]
> > ```js
> > doSomething();  // Maybe ends the task, mabye doesn't.
> > ```
> 
> Synchronous functions don't end a task, because JS is run to completion. Did you mean `await doSomething()` ?

You've hit the nail on the head. `doSomething()` **is in fact** asynchronous, even if you don't `await` on it. It waits on a Promise. Sometimes it waits on an already-resolved Promise, sometimes it waits for 5s.

```js
async function doSomething() {
  await (makeDecision() ? alreadyResolvedPromise : slowPromise);
}
```

Now, it's exactly [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) which I am citing here:

```js
const stream1 = await navigator.mediaDevices.getDisplayMedia();
doSomething();  // Maybe ends the task, maybe doesn't.
navigator.mediaDevices.focusPolicy = "no-focus";
```

This code produces flaky results, but **cannot raise exceptions**.
That's not simple.
That's a problem.

**Comment by jan-ivar** [comment]
> `doSomething()` **is in fact** asynchronous, even if you don't `await` on it. It waits on a Promise.

@eladalon1983 But the caller doesn't (a)wait for it. Please check your assumptions: https://jsfiddle.net/jib1/0sjgd6cb/

**Comment by jan-ivar** [comment]
> Each then returns a new promise, but this is fine. The idea is for the microtask used by User Agent to check for focus-or-not to be executed within the same event loop task but after the JS shim getDisplayMedia promise resolution callbacks. I think this is doable with the updated shim I provided, can you explain why it would not work? I tried to express my understanding/POC in https://jsfiddle.net/f0uqx8jg/.

@youennf Your fiddle is more lax than what the [OP proposes](https://wicg.github.io/conditional-focus/#dom-focusablemediastreamtrack-focus): _"MT finishes"_, which is what I was countering. But you've shown that the OP [isn't actually implementable](https://jsfiddle.net/jib1/04oL27c5/), so thanks for that.

But I disagree shimming should be this hard, even with your proposal. Shims should be able to return a different promise due to synchronous post-success code. And I see no need for it, because you've not shown any benefit over simply queening a task.

**Comment by youennf** [comment]
> @youennf Your fiddle is more lax than what the [OP proposes](https://wicg.github.io/conditional-focus/#dom-focusablemediastreamtrack-focus): _"MT finishes"_

That might be true, I haven't looked at Elad's spec.
I think Elad and I are mostly aligned there on this 'lax' proposal, @eladalon1983?.

> I disagree shimming should be this hard

Can you describe how hard that is? Shiming is a subtle job and I'd like to understand which getDisplayMedia shiming use cases we are talking about to evaluate whether this is impossible, just hard or just subtle.

> Shims should be able to return a different promise due to synchronous post-success code.

In that case, it becomes very subtle whether calling await will make calling the focus API working or not.
I am favouring web developer simplicity over shiming power/shiming authoring.

> I see no need for it, because you've not shown any benefit over simply queening a task.

I think we covered some benefits already but here are some:
- Simplicity: you cannot call await, instead you need to decide synchronously within one of getDisplayMedia promise registered resolution callback.
- Earlier is better than later when it comes to let UA decide whether to focus or not. By scheduling another task, the risk is to further delay the decision. As I said, fetch event decided for a synchronous constraint.
- After the micro tasks queue is emptied, there are some additional work to be done to finish the current task (contextrestored, animation frame callback...), which will further delay execution of the next task.
- An event loop has several task queues. There is no guarantee which task queue will be picked in the next event loop, so it is difficult to guarantee that the next task will actually be the task to let UA decide focus. This might lead to potential flakinesses or subtle interop issues.

My understanding is that you would like to be able to run the Microtask to select UA decision once all JS-observable microtasks have run. So far, we do not have that infrastructure and the closest to that is queue a microtask in the gym promise resolution task, not queue a task.

**Comment by jan-ivar** [comment]
I think you're confusing simplicity with strictness.

Developers expect to be allowed to split code into sub-functions and add logging statements. Some devs even seem to like to structure their code this way for no reason at all. I personally don't, but it MUST continue to work:
```js
// My first screen sharing app!

// The screen sharing function. I'm excited to learn this! I hope it works...
async function getDisplayMedia(constraints) {
  try {
    const stream = await navigator.mediaDevices.getDisplayMedia(constraints);
    console.log(`Debug: screen-share succeeded. Got ${stream.getTracks().length} track(s)`);
    return stream;
  } catch (e) {
    console.log(`Debug: screen-share failed ${e.message}`);
    throw e;
  }
}

button1.onclick = async () => {
  try {
    video.srcObject = await getDisplayMedia({video: true});
    navigator.mediaDevices.displayMediaFocusMode = "no-focus"; // why no worky??
  } catch(e) {
    console.log(e);
  }
};

button2.onclick = async () => {
  try {
    video.srcObject = await navigator.mediaDevices.getDisplayMedia({video: true});
    navigator.mediaDevices.displayMediaFocusMode = "no-focus"; // works now! Thanks stackoverflow!
  } catch(e) {
    console.log(e);
  }
};
```
Your proposal breaks this fundamental expectation. https://jsfiddle.net/jib1/y5rz3hpm/

**Comment by jan-ivar** [comment]
> By scheduling another task, the risk is to further delay the decision.

Why are we in fear of queuing a task? Have you measured this?

> This might lead to potential flakinesses or subtle interop issues.

What evidence do you have of this?

> My understanding is that you would like to be able to run the Microtask to select UA decision once all JS-observable microtasks have run. So far, we do not have that infrastructure and the closest to that is queue a microtask in the gym promise resolution task, not queue a task.

No I want to queue a task. We queue tasks all the time, and we've shown here we can queue tasks A and B to happen in close succession. I'm not buying any issue of risk of delay.

**Comment by youennf** [comment]
> > By scheduling another task, the risk is to further delay the decision.
> 
> Why are we in fear of queuing a task? Have you measured this?

As per spec, additional events or JS callbacks are potentially called, plus potentially other tasks in between the two tasks.
The more we call web app JS, the more delay there is.
Getting a synchronous decision in that particular case seems good, just like being able to prevent default actions for input events is happening synchronously.

> > This might lead to potential flakinesses or subtle interop issues.
> 
> What evidence do you have of this?

Spec says in https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model.
`Let taskQueue be one of the event loop's task queues, chosen in an implementation-defined manner, with the constraint that the chosen task queue must contain at least one runnable task. If there is no such task queue, then jump to the microtasks step below`

`For example, a user agent could have one task queue for mouse and key events (to which the user interaction task source is associated), and another to which all other task sources are associated. Then, using the freedom granted in the initial step of the event loop processing model, it could give keyboard and mouse events preference over other tasks three-quarters of the time, keeping the interface responsive but not starving other task queues.`

> No I want to queue a task.

Do you have a good example of an existing API where we queue a task in a similar scenario as this one where UA needs to wait for some JS to do a default UI action?

> No I want to queue a task.

Why is queuing a task better than the specific Microtask approach I mentioned?

**Comment by jan-ivar** [comment]
Let's measure! https://jsfiddle.net/jib1/y1e4rg0d/ (Firefox result on a 2019 MBP, other browsers are similar 3K-4K)
```
tasks / setTimeout(100ms) = 4655
tasks / ms = 45.64
```
Transient activation in all browsers I tested: https://jsfiddle.net/jib1/zqxnu9t5/ : >4 seconds (or >182,560 tasks)

So according to browsers, >4 seconds is when a user stops associating a reaction with their action.

So from a security perspective, we're fine by several orders of magnitude. From a general UX snappiness perspective, we're also fine by an order of magnitude.

> The more we call web app JS, the more delay there is

That's a quantitative claim of linearity I don't think holds or is helpful. By that rationale, tightening our purse strings on tasks should improve performance. But it doesn't. A single microtask can wait forever. It's nonlinear. I don't think there can be any inherent correlation between an amount of work being done and the number of tasks queued. It would depend entirely on how an application is structured.

This seems like a nonissue to me.

**Comment by youennf** [comment]
> So according to browsers, >4 seconds is when a user stops associating a reaction with their action.

We are talking about reducing it to 1 second as a security mitigation, probably counted in another process so we should probably reduce to half of it in the process running JS.
And we want to run in slow devices, with loaded CPU and in low power mode.
And browsers might want to sneak a few input events between the two tasks, especially when CPU is loaded.
And the whole thing with good interoperability and predictability across all browser implementations.

Let's summarise the four proposals:
1. queue a microtask: ready to spec, consistent behavior but use of await would not be possible (seen as either a pro or a con).
2. queue a task: ready to spec, but potential added delay and potential interop issues, some use of await possible (seen either as a pro or con).
3. queue a microtask that is running after the JS registered microtasks: not spec ready, consistent behavior, some use of await possible.
4. queue a task that is guaranteed to run next to the resolution task: not spec ready, would remove the potential interop issues and some of the delay from option 2.

**Comment by jan-ivar** [comment]
On Firefox for Android on my Samsung S9 I get
```
tasks / setTimeout(100ms) = 1301
tasks / ms = 12.88
```
That's 12 tasks per _millisecond_ on a 3½ year-old phone. The math doesn't seem to add up to anything UX-observable. Not even close.

Also you haven't addressed the problems I raised. The only proposal that doesn't break [fundamental expectations](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-929491274) are options 2 and 4 (2 seems realistic to me).

We're talking about delaying when browsers focus a window by 83 microseconds average from the time the user accepts a prompt, so JS can have a chance to override focus.

> ... browsers might want to sneak a few input events between the two tasks, especially when CPU is loaded.

If you consider this an interoperability issue, you should raise an issue on the JS event loop model. cc @domenic

**Comment by eladalon1983** [comment]
Your measurement does not attempt to do anything CPU-intensive. Applications might.

**Comment by eladalon1983** [comment]
I am trying to **reformulate** my position more clearly. Hopefully this won't be mere repetition. If it is - my apologies.

It might be worth it **working backwards** here:

Issue 1, [WoO](https://en.wikipedia.org/wiki/Window_of_opportunity)-closing-mechanism.
* We have NOT set out to allow applications to change focus to a captured surface at an arbitrary time.
* We don't want to allow applications to change focus any later than a "reasonable" time after capturing starts.
* All mechanisms discussed thus far (task/microtask) can be gamed through busy-waiting.
* For that reason, **besides** any "happy-case" mechanism (task/microtask), **additional** enforcement by the UA is necessary in the form of a maximum time-period.
* This cannot be the only mechanism, because it would not be snappy in the _legacy_ happy-case (tasks/microtasks are fast but the app does not call the API).
* Because of CPU scheduling issues, even optimal applications will **sometimes** hit this limit. When that happens, it's non-actionable and requires no exception. (Also, an exception here is hard to implement, as the time-limit may be imposed in another process for security reasons - particularly in Chrome's implementation.)
* A good API will minimize the number of times this issue is encountered, by encouraging the application to do the minimal amount of work between the time the Promise is resolved and the time it calls the API.
* Limiting the "happy-case" mechanism on microtask boundary rather than task boundary pushes the application harder in this desirable direction.

Issue 2, future-facing APIs:
* Applications can have branching logic in between the Promise being resolved and the API being triggered.
* This means an application could find that the API sometimes works, sometimes doesn't - for example, if it relies on data that it might-or-might-not already have in its app-level cache. (E.g. they might have previously fetched some data from a remote source, and it might already have been downloaded or might not have.)
* It's important to fail loudly and clearly in these cases.
* An API that's valid to call in order to influence the future, cannot raise exceptions when it's called too late to influence the past.
* Therefore, the global-attribute-influencing-future API cannot be our solution.
* It may **complement** our solution, though, by setting a **default** for future calls.

**Comment by annevk** [comment]
I wanted to clarify something about the fetch event and also ask a question.

The way fetch events work is basically:

1. Spec dispatches a fetch event.
2. This runs event listeners. As the user agent dispatched the event event, the microtask queue is drained here as well. (No JavaScript on the stack has that effect.)
3. Spec inspects state on the event.

This is a fairly normal pattern for event dispatching.

The problem with 2 in https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-929625605 (queue a task) is that user interaction events might get in between and end up changing what the user wants to do. 1 seems rather surprising for web developers. At least the idea behind promises is that you can chain them a bunch with `then()`, even for what are normally synchronous tasks. This would end up invalidating that, which would be rather surprising I think.

My question is, did you consider passing a callback to `getUserMedia()` or dispatching an event to take care of this instead? That would give you a time-bound period that allows for custom logic and doesn't run into some of the issues you are facing. Failing that, I think 3 from @youennf's list could be an option. This would essentially require adding a step around 5/6 of https://html.spec.whatwg.org/#perform-a-microtask-checkpoint. It seems less elegant though.

Hope that helps.

**Comment by eladalon1983** [comment]
> did you consider passing a callback to getUserMedia() or dispatching an event to take care of this instead?

(Assuming you mean `getDisplayMedia`.)

I have considered that. [1]

If the callback is executed **before** resolving the Promise, we have a way for the application to interact with an MediaStream before the Promise that returns it is resolved. This seems like a footgun to me; now gDM might return a Promise to a MediaStream that's had its video track removed, or a new video track inserted. Oy.

If the callback is executed **after** resolving the Promise, then when exactly? As a scheduled next microtask? Task? We end up in a new version of the same discussion, and we leave footguns lying around with intervening tasks/microtasks that could end up taking too long to execute, closing the window on account of the 1s (or other value) limit.

--
[1] Edit: Considered a callback, I mean. I don't think an event is ergonomic, and I think many of the problems discussed above still apply to an event.

**Comment by annevk** [comment]
`getDisplayMedia()` at some point queues a task to resolve the promise. That's where you'd execute the callback. This would end up happening before the `getDisplayMedia().then()` callback, which would happen during the microtask portion of that task. As it's all in the same task asked for by a single caller I don't think it's a footgun, but I might be misunderstanding your point.

**Comment by eladalon1983** [comment]
So you're proposing accepting a callback and scheduling two microtasks back-to-back, one for resolving the Promise and one to run [after](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-932160209) the first. I think this is quite similar to the two-microtasks-back-to-back approach without a callback, while also making it less clear to Web-developers that the callback could "expire" if the first microtask is delayed for too long.

**Comment by annevk** [comment]
It would be more like this:

1. Queue a task to:
   1. Call the callback.
   2. Lock focus.
   3. Resolve the promise.
 
(Where you place step 1.3 doesn't really matter as it'll only be observable after 1.1 and 1.2)

**Comment by eladalon1983** [comment]
I am not sure I understand your proposal. It seems like the one I classified as "before" in an [earlier](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-932160209) comment, but I could be misunderstanding.

```js
function callback(mediaStream) {
  console.log('b');
  mediaStream.removeTrack(mediaStream.getVideoTracks()[0]);
  console.log('c');
}

console.log('a');
const mediaStream = navigator.mediaDevices.getDisplayMedia({video: true}, callback);
console.log('d');
const [track] = (await mediaStream).getVideoTracks();
console.log('e');
```

What's going to happen with the code above?

**Comment by annevk** [comment]
Order would be a, d, b, c, e if I'm not mistaken. You can indeed remove tracks, but this can also happen if you hand out the promise in multiple places. That doesn't really seem problematic to me.

**Comment by eladalon1983** [comment]
So you were indeed referring to the **before** option from [this comment](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-932160209), which means I was right to edit away my misunderstanding in ~[this comment](https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-932186120)~. Glad it's clarified. :-)

I think giving access to the MediaStream to something that executes before the Promise resolves violates [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). Are you aware of precedents, perhaps?

**Comment by annevk** [comment]
Yeah, e.g., https://notifications.spec.whatwg.org/#dom-notification-requestpermission. I'm pretty sure we also have places that dispatch an event and resolve a promise in the same task. Stems from how promises work. (And to be pedantic, the promise can be resolved before, but it being resolved cannot be observed until after.)

**Comment by jan-ivar** [comment]
How about this:
```js
const stream = await navigator.mediaDevices.getDisplayMedia();
navigator.mediaDevices.addEventListener("focus", e => decide(stream) || e.preventDefault(), {once: true});
```
This would be the entire API surface. No need for exceptions. I'm open to bikeshedding the event name.

**Comment by eladalon1983** [comment]
> How about this:
> 
> ```js
> const stream = await navigator.mediaDevices.getDisplayMedia();
> navigator.mediaDevices.addEventListener("focus", e => decide(stream) || e.preventDefault(), {once: true});
> ```
> 
> This would be the entire API surface. No need for exceptions. I'm open to bikeshedding the event name.

The use of `preventDefault` suggests that the default behavior, hitherto unspecified, would now be specified as focus-captured-surface. Is that right?

**Comment by jan-ivar** [comment]
Yes.

**Comment by annevk** [comment]
FWIW, as I understand it the task responsible for resolving `getDisplayMedia()` has been traditionally also the focus task. It seems to me that therefore you would have to expose the stream on the focus event. Delaying the focus by a task in the hope that the caller of `getDisplayMedia()` holds a stream at that point still seems rather unclean.

**Comment by eladalon1983** [comment]
I'm wondering if we could in the future run into an issue where some operating systems require a separate permission before they allow the UA to switch focus between applications. We might want to get around that by specifying the default behavior as "the user agent MUST try to focus the captured surface." Wdyt?

**Comment by youennf** [comment]
I would also prefer just one task.
My understanding of the algorithm would be something like: queue a task that runs the following steps:
- fire a focus event (taking stream, a subset of it, stream's clone...)
- execute focus algorithm based on focus policy computed from focus event handling. 
- resolve getDisplayMedia promise with stream

I do not see why we should delay the focus algorithm.

**Comment by jan-ivar** [comment]
> FWIW, as I understand it the task responsible for resolving getDisplayMedia() has been traditionally also the focus task.

I ran a test to disprove this https://jsfiddle.net/jib1/q75yb8pf/ but I was also surprised to learn that Firefox blurs the capturer window 15 ms _before_ it resolves the promise (I thought it happened much later):
```
4823: blurred
4837: resolved
```
But if we're worried about web compat here we have a much bigger problem: Chrome's prompt steals focus.
```
1827: blurred
16078: resolved
```
We've never specified exactly when (non-browser) window is focused before, and I'm not sure we need to here. What we're specifying here is a JS control opportunity by requiring the browser to fire a warning ping _"hey, I plan to focus another window, you ok with that?"_, which to me doesn't inherently need to match up with any user-visible behavior timing wise.

**Comment by jan-ivar** [comment]
> I would also prefer just one task. My understanding of the algorithm would be something like: queue a task that runs the following steps:
> 
>  * fire a focus event (taking stream, a subset of it, stream's clone...)

I don't see a problem with returning the original. Clones are trouble.

>  * execute focus algorithm based on focus policy computed from focus event handling.

I'd call it the "focus decision algorithm", since it doesn't seem necessary to specify when exactly actual focusing takes place.

>  * resolve getDisplayMedia promise with stream

Let's say I did this:
```js
async function getDisplayMedia() {
  const stream = await navigator.mediaDevices.getDisplayMedia();
  console.log("A");
  return stream;
}

const p = getDisplayMedia();
await new Promise(r => navigator.mediaDevices.onfocus = e => r(decide(e.stream) || e.preventDefault()));
console.log("B");
const stream = await p;
```
Would this produce `A B` or `B A`? I'm not sure microtask order here is even standardized.

> I do not see why we should delay the focus algorithm.

Tasks have cleaner semantics than messing with microtask order.

**Comment by annevk** [comment]
It would produce B, A. As there's no JavaScript on the stack, the microtask is queue is drained after event callback invocation. I think that indeed makes the order observable and therefore it's good to use the order @youennf proposed.

**Comment by youennf** [comment]
Another approach to consider is to use a callback given to getDisplayMedia as input parameter.
It does not allow to set a context wide behavior but it makes it very easy to understand to which getDisplayMedia call the decision should be made.

**Comment by happylinks** [comment]
Just wanted to give my vote for having this API again. 
Today yet another user of our application told us they thought they were already recording their screen after they shared their tab and it jumped to the tab they shared. 
We already try to hint them to come back to our tab through the title and favicon like this: 
![image](https://user-images.githubusercontent.com/184774/162440494-39548bd6-a1b9-436f-a4a4-2426d8a44513.png)
And we try to explain that they should always come back to our tab in our UI, but still users get confused sometimes.

Because we record two streams at the same time, and we need to sync those up, starting recording automatically on a screen share is not good enough. MediaRecorders take time to start, sometimes even multiple seconds. We need to let the user know exactly when the recorder started so they don't start talking before we are actually recording.

In our application specifically we never want the auto-focus behavior to happen and would love to be able to disable it.
